finality-aleph/src/sync/handler/mod.rs

```diff
@@ -679,56 +716,35 @@ where
                     if self.forest.skippable(&h.id()) {
                         continue;
                     }
-                    let h = match self.verify_header(h, false) {
-                        Ok(VerifiedHeader {
-                            header: h,
-                            maybe_equivocation_proof: Some(proof),
-                        }) => {
-                            equivocation_proofs.push(proof);
-                            h
-                        }
-                        Ok(VerifiedHeader {
-                            header: h,
-                            maybe_equivocation_proof: None,
-                        }) => h,
-                        Err(e) => return (new_highest, equivocation_proofs, Some(e)),
-                    };
-                    if let Err(e) = self.forest.update_header(&h, Some(peer.clone()), false) {
-                        return (new_highest, equivocation_proofs, Some(Error::Forest(e)));
+                    if let Err(e) =
+                        self.handle_header(h, peer.clone(), &mut equivocation_proofs, None)
+                    {
+                        return (new_highest, equivocation_proofs, Some(e));
+                    }
+                }
+                ResponseItem::Block(b) => {
+                    if self.forest.skippable(&b.header().id()) {
+                        continue;
                     }
-                    if !self.forest.importable(&h.id()) {
+                    if let Err(e) = self.handle_header(
+                        b.header().clone(),
+                        peer.clone(),
+                        &mut equivocation_proofs,
+                        last_imported,
+                    ) {
                         return (
                             new_highest,
                             equivocation_proofs,
-                            Some(Error::HeaderNotRequired(h.id())),
+                            Some(match e {
+                                Error::HeaderNotRequired(id) => Error::BlockNotImportable(id),
+                                _ => e,
+                            }),
                         );
                     }
-                }
-                ResponseItem::Block(b) => {
-                    if self.forest.skippable(&b.header().id()) {
-                        continue;
+                    last_imported = Some(b.header().id());
+                    if let Err(e) = self.import_block(b, false) {
```

Comment:
`import_block` here now completely unnecessarily verifies the header and checks for equivocations, please fix that – keeping in mind that we still have to do that for own blocks. Perhaps rename to `import_own_block` and here use `block_importer.import_block` directly?

Reply:
Good point. I noticed that but wanted to avoid changing logic too much. Changed as suggested.